README
Devel-DumpTrace
---------------

A debugging module that displays the statements being executed,
substituting variable identifiers with their values.

Example:

    $ cat demo.pl
    #!/usr/bin/perl
    $a = 1;
    $b = 3;
    $c = 2 * $a + 7 * $b;
    @d = ($a, $b, $c + $b);

    $ perl -d:DumpTrace=verbose demo.pl
    >>  demo.pl:2:[__top__]:
    >>>              $a = 1;
    >>>>>            1 = 1;
    -------------------------------------------
    >>  demo.pl:3:[__top__]:
    >>>              $b = 3;
    >>>>>            3 = 3;
    -------------------------------------------
    >>  demo.pl:4:[__top__]:
    >>>              $c = 2 * $a + 7 * $b;
    >>>>             $c = 2 * 1 + 7 * 3;
    >>>>>            23 = 2 * 1 + 7 * 3;
    -------------------------------------------
    >>  demo.pl:5:[__top__]:
    >>>              @d = ($a, $b, $c + $b);
    >>>>             @d = (1, 3, 23 + 3);
    >>>>>            (1,3,26) = (1, 3, 23 + 3);
    -------------------------------------------

INSTALLATION

To install this module, follow the usual recipe:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Devel::DumpTrace

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Devel-DumpTrace

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Devel-DumpTrace

    CPAN Ratings
        http://cpanratings.perl.org/d/Devel-DumpTrace

    Search CPAN
        http://search.cpan.org/dist/Devel-DumpTrace/


LICENSE AND COPYRIGHT

Copyright (C) 2010-2013 Marty O'Brien

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Devel-DumpTrace

0.18    2013-03-27
	Accomodate changes in hash order in Perl >v5.17.10

0.17    2012-06-07
	Use Hash::SafeKeys as permanent solution to RT#77673.

0.16    2012-06-06
	RT#77673 infinite loop with PPI and each %hash construction, fixed
	    using Storable::dclone to copy hash without resetting iterator
	Check,trap whether t/89-incl-exl.t is affected by Perl v5.8.8 bug with
	    calls like  "perl -d:Foo=arg ..."

0.15    2011-11-28
	Put PPI into separate Devel::DumpTrace::PPI package.
	Put useful constants into separate Devel/DumpTrace/Const.pm file
	Support for DUMPTRACE_COUNT=true to count how many times each line
	    was executed.
	Fixed bug that prevented DUMPTRACE_TIME from using Time::HiRes::time
	Fixed deep recursion issue in Devel::DumpTrace::dump_scalar

0.14    2011-07-14
	Fix for t/ppi/33.t failure on some systems.

0.13    2011-07-12
	Evaluate subscript arguments and provide smarter abbreviations
	    of large arrays and hashes.
	Upgrade bundled Text::Shorten to 0.04.
	Detect when implicit $_~~ is used in some given/when blocks.

0.12    2011-03-22
	Support for DUMPTRACE_PID and DUMPTRACE_TIME env var.
	Support for DUMPTRACE_FH=tty.
	Support for DUMPTRACE=a=b,c=d,... as shorthand for 
	    DUMPTRACE_a=b,DUMPTRACE_c=d
	Fix inappropriate implicit $_ on barewords that resemble builtin names.
	Fix deferred evaluation of complex lvalues.

0.11    2011-02-25
	Support to include/exclude packages at import. This also fixes the
	    uninitialized hash warning with v<=5.8.

0.10    2011-02-18
	Implemented decorator (see 0.09) for for[each] LIST construction.
	Implemented decorator for do-while and do-until constructions.
	Fixed PPI parsing to treat source code specified from -e, eval, or
	    standard input the same as code from a file.
	Fixed bug about inappropriate implicit $_ inserted after builtins.

0.09    2011-02-16
	Defer output of statements containing function calls until after the
	    function call completes.
	Refresh pads to evaluate lexical where assignment and declaration are
	    in the same statement. That is, we now get  my $x=7 --> my $x:7=7
	    and not my $x=7 --> my $x:undef=7.
	Implemented "decorators" for C-style for loops, while/until loops,
	    if/elsif/else statements to show expressions that were implicitly
	    evaluated between potential breakpoints in the trace output.
	Include subroutine name in trace output.

0.08    2011-02-10
	More efficiency improvements for abbreviating large arrays and hashes.
	Fix in Text/Shorten.pm that for code that was incompatible with
	    Perl v5.8.
	More PPI tweaks to improve displayed results.

0.07    2011-02-08
	Rename $XTRACE_FH to $DUMPTRACE_FH and XTRACE_xxx env vars to 
	    DUMPTRACE_xxx.
	PPI tweaks to improve displayed results.
	Made Devel::DumpTrace::PPI the default module (when PPI is available)
	    and added Devel::DumpTrace::noPPI to explicitly use non-PPI vers.
	Efficiency improvements for abbreviating large arrays and hashes.

0.06    2011-02-01
	Rename distribution from Devel-XTrace to Devel-DumpTrace.
	Release candidate.

0.05    2011-01-31
	PPI fixes. Added smoke tests. 
	PPI detects when implicit $_,@_,@ARGV should be inserted.

0.04    2010-11-30
	Initial PPI implementation.
	Rename $LEVEL to $TRACE and aliased to $Devel::Trace::TRACE.

0.03    2010-11-26
	Implement output reduction strategies and $LEVEL verbosity level var.

0.02    2010-11-15
	Remove some of those limitations from v0.01.
	- Handle package vars
	- Handle special vars $ARGV, %SIG, $1, etc.

0.01    2010-11-05
	Initial version. Many "BUGS AND LIMITATIONS"


t/89-incl-excl-pkgs.t.pl

{
  package Abc::Def::Ghi;

  sub jkl {
    42;
  }

}

{
  package Abc::Ghi::Jkl;

  sub mno {
    19;
  }

}

package Abc::Def;
Abc::Def::Ghi::jkl();
Abc::Ghi::Jkl::mno();



lib/binary_search.pl
# This file demonstrates a use case of Devel::DumpTrace:
# debugging a binary search function.
# Eventually:
#    the other documentation will refer to this file
#    there will be another file with the fixed function




# ...
# so we have written this binary search function:

sub binary_search {
  my ($target, @list) = @_;
  my ($lo,$hi,$mid,$w) = (0,$#list);
  while ($hi - $lo > 1) {
    $mid = int(($lo+$hi)/2);
    $w = $list[$mid];
    if ($target lt $w) {
      $hi = $mid - 1;
    } else {
      $lo = $mid + 1
    }
  }
  if ($list[$lo] ge $target) {
    return $lo;
  } else {
    return $hi;
  }
}

$Devel::DumpTrace::LEVEL = 0;

# at first glance it looks correct but it fails on some test cases

@w = sort qw(the quick brown fox jumps over the lazy dog);
for $w (@w) {
  $ww = binary_search($w, @w);
  if ($w[$ww] eq $w) { 
    print "$ww $w OK\n";
  } else { 
    print "$ww $w FAIL\n";
  }
}

# for $w := 'lazy', the return value is 5 but it should be 4.
# Why did this fail?
# Let's debug it with Devel::DumpTrace

BEGIN { $Devel::DumpTrace::TRACE = 0 };


$Devel::DumpTrace::TRACE = 'verbose';
  
$ww = binary_search($w[4], @w);
print "\$ww is $ww, should be 4.\n";

$Devel::DumpTrade::TRACE = 0;

# Examine the output. Can you spot the error? Can you fix it?




lib/Devel/DumpTrace/Const.pm
package Devel::DumpTrace::Const;
use Exporter;
use strict;
use warnings;

our $VERSION = '0.18';
our @ISA = qw(Exporter);
our @EXPORT = qw/
    DISPLAY_NONE DISPLAY_TERSE DISPLAY_GABBY
    ABBREV_SMART ABBREV_STRONG ABBREV_MILD_SM ABBREV_MILD ABBREV_NONE
    OUTPUT_SUB OUTPUT_PID OUTPUT_TIME OUTPUT_COUNT
    CALLER_PKG CALLER_SUB
/;

# parameters for controlling how much output is produced
use constant DISPLAY_NONE  => 0;  # trace off
use constant DISPLAY_TERSE => 1;  # concise - 1 trace line per stmnt
use constant DISPLAY_GABBY => 4;  # verbose - 2-5 trace lines per stmt
use constant ABBREV_SMART  => 0;  # strong,smart abbrev of long scalars,
use constant ABBREV_STRONG => 1;  # strong abbreviation of long scalars,
use constant ABBREV_MILD_SM => 2; # mild abbreviation      arrays, hashes
use constant ABBREV_MILD   => 3;  # mild abbreviation      arrays, hashes
use constant ABBREV_NONE   => 4;  # no abbreviation

# additional information to include in output
use constant OUTPUT_SUB => !($ENV{DUMPTRACE_NO_SUB} || 0) || 0;
use constant OUTPUT_PID => $ENV{DUMPTRACE_PID} || 0;
use constant OUTPUT_TIME => $ENV{DUMPTRACE_TIME} || 0;
use constant OUTPUT_COUNT => $ENV{DUMPTRACE_COUNT} || 0;

# for interpreting list output of  caller
use constant CALLER_PKG => 0;     # package name
use constant CALLER_SUB => 3;     # current subroutine name

1;


lib/Devel/DumpTrace/CachedDisplayedHash.pm
#
# tied hash object that maintains two additional states:
#   1. a parallel hash table where all original keys and values
#      are run through  &Devel::DumpTrade::dump_scalar
#   2. a cache of results from  Text::Shorten::shorten_hash
#
# Calls to  Devel::DumpTrace::hash_repr  should not trigger calls
# to  Text::Shorten::shorten_hash  unless the hash table
# has been updated since the last  shorten_hash  call.
#

package Devel::DumpTrace::CachedDisplayedHash;

use strict;
use warnings;
use Carp;
our $VERSION = '0.18';

*dump_scalar = \&Devel::DumpTrace::dump_scalar;


sub TIEHASH {
  my ($class, @list) = @_;

  # HASH:  the original and primary hash table

  # PHASH: copy of HASH where all keys and values are
  #        filtered through Devel::DumpTrace::dump_scalar;

  # CACHE: store of results from Text::Shorten. Keys are
  #        auxiliary arguments to Text::Shorten::shorten_hash,
  #        values are array refs of shorten_hash return values.
  #        Cache is cleared when any element of the hash
  #        is changed.

  my $self = {
	      CACHE => {},
	      HASH => { @list },
	      PHASH => { map { dump_scalar($_) } @list }
	     };
  return bless $self, $class;
}

sub FETCH {
  my ($self, $key) = @_;
  return $self->{HASH}{$key};
}

sub STORE {
  my ($self, $key, $value) = @_;
  $self->clear_cache;
  my $old = $self->{HASH}{$key};
  $self->{HASH}{$key} = $value;
  $self->{PHASH}{dump_scalar($key)} = dump_scalar($value);
  return $old;
}

sub DELETE {
  my ($self, $key) = @_;
  $self->clear_cache;
  delete $self->{PHASH}{dump_scalar($key)};
  return delete $self->{HASH}{$key};
}

sub CLEAR {
  my $self = shift;
  $self->clear_cache;
  $self->{PHASH} = {};
  $self->{HASH} = {};
  return;
}

sub EXISTS {
  my ($self, $key) = @_;
  return exists $self->{HASH}{$key};
}

sub FIRSTKEY {
  my $self = shift;
  scalar keys %{$self->{HASH}};
  return each %{$self->{HASH}};
}

sub NEXTKEY {
  my ($self, $lastkey) = @_;
  return each %{$self->{HASH}};
}

sub SCALAR {
  my $self = shift;
  return scalar %{$self->{HASH}};
}

# sub UNTIE { } # not implemented
# sub DESTROY { } # not implemented

sub clear_cache {
  my $self = shift;
  $self->{CACHE} = {};
  return;
}

sub store_cache {
  my ($self, $key, $value) = @_;
  $self->{CACHE}{$key} = $value;
  if ($Devel::DumpTrace::HASHREPR_SORT) {
      my %h = @$value;
      $value = [ map { $_ => $h{$_} } sort keys %h ];
  }
  return;
}

sub get_cache {
  my ($self, $key) = @_;
  return $self->{CACHE}{$key};
}

1;


lib/Devel/DumpTrace/CachedDisplayedArray.pm
#
# tied array object that maintains two additional states:
#   1. a parallel array where all original entries
#      are run through  &Devel::DumpTrade::dump_scalar
#   2. a cache of results from  Text::Shorten::shorten_hash
#
# Calls to  Devel::DumpTrace::array_repr  should not trigger calls
# to  Text::Shorten::shorten_array  unless the array
# has been updated since the last  shorten_array  call.
#

package Devel::DumpTrace::CachedDisplayedArray;

use strict;
use warnings;
use Carp;
our $VERSION = '0.18';

*dump_scalar = \&Devel::DumpTrace::dump_scalar;


sub TIEARRAY {
  my ($class, @list) = @_;

  # ARRAY:  the original and primary hash table

  # PARRAY: copy of HASH where all keys and values are
  #         filtered through Devel::DumpTrace::dump_scalar;

  # CACHE:  store of results from Text::Shorten. Keys are
  #         auxiliary arguments to Text::Shorten::shorten_array,
  #         values are array refs of shorten_array return values.
  #         Cache is cleared when any element of the array
  #         is changed.

  my $self = {
	      CACHE => {},
	      ARRAY => [ @list ],
	      PARRAY => [ map { dump_scalar($_) } @list ]
	     };
  return bless $self, $class;
}

sub FETCH {
  my ($self, $index) = @_;
  return $self->{ARRAY}[$index];
}

sub STORE {
  my ($self, $index, $value) = @_;
  $self->clear_cache;
  my $old = $self->{ARRAY}[$index];
  $self->{ARRAY}[$index] = $value;
  $self->{PARRAY}[$index] = dump_scalar($value);
  return $old;
}

sub FETCHSIZE {
  my $self = shift;
  return scalar @{$self->{ARRAY}};
}

sub STORESIZE {
  my ($self, $newcount) = @_;
  my $oldcount = $self->FETCHSIZE();
  if ($newcount > $oldcount) {
    $self->clear_cache;
    $self->STORE($_, undef) for $oldcount .. $newcount-1;
  } elsif ($newcount < $oldcount) {
    $self->clear_cache;
    $self->POP() for $newcount .. $oldcount-1;
  }
  return;
}

sub EXTEND {
  return;
}

sub DELETE {
  my ($self, $index) = @_;
  $self->clear_cache;
  return $self->STORE($index, undef);
}

sub CLEAR {
  my $self = shift;
  $self->clear_cache;
  $self->{PARRAY} = [];
  $self->{ARRAY} = [];
  return;
}

sub EXISTS {
  my ($self, $index) = @_;
  return exists $self->{ARRAY}[$index];
}

sub PUSH {
  my ($self, @list) = @_;
  if (@list > 0) {
    $self->clear_cache;
  }
  push @{$self->{ARRAY}}, @list;
  push @{$self->{PARRAY}}, map { dump_scalar($_) } @list;
  return $self->FETCHSIZE();
}

sub POP {
  my $self = shift;
  if (@{$self->{ARRAY}} > 0) {
    $self->clear_cache;
  }
  pop @{$self->{PARRAY}};
  return pop @{$self->{ARRAY}};
}

sub SHIFT {
  my $self = shift;
  if (@{$self->{ARRAY}} > 0) {
    $self->clear_cache;
  }
  shift @{$self->{PARRAY}};
  return shift @{$self->{ARRAY}};
}

sub UNSHIFT {
  my ($self, @list) = @_;
  if (@list > 0) {
    $self->clear_cache;
  }
  unshift @{$self->{PARRAY}}, map { dump_scalar($_) } @list;
  my $result = unshift @{$self->{ARRAY}}, @list;
  return $result;
}

sub SPLICE {
  my ($self, $offset, $length, @list) = @_;
  $offset ||= 0;
  $length ||= $self->FETCHSIZE() - $offset;

  $self->clear_cache;
  splice @{$self->{PARRAY}}, $offset, $length, map { dump_scalar($_) } @list;
  return splice @{$self->{ARRAY}}, $offset, $length, @list;
}

# sub UNTIE { } # not implemented
# sub DESTROY { } # not implemented

sub clear_cache {
  my $self = shift;
  $self->{CACHE} = {};
  return;
}

sub store_cache {
  my ($self, $key, $value) = @_;
  $self->{CACHE}{$key} = $value;
  return;
}

sub get_cache {
  my ($self, $key) = @_;
  return $self->{CACHE}{$key};
}

1;


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;
use v5.8.2;               # PadWalker 1.5 not available for <v5.8.2

my $TESTS = "t/*.t t/ppi/*.t";
unless (eval 'use PPI;1') {
  print q[
This distribution comes with a PPI-based module.
Install the PPI distribution on your system
to make use of it.

];
  $TESTS = "t/*.t t/ppi/00-load.t";
}

# no index instructions for PAUSE
my $NO_INDEX = 
  { 'package' =>
    ['Text::Shorten', 
     'Devel::DumpTrace::CachedDisplayedArray',
     'Devel::DumpTrace::CachedDisplayedHash', 
     'Devel::DumpTrace::noPPI'] };

WriteMakefile(
    NAME                => 'Devel::DumpTrace',
    AUTHOR              => q{Marty O'Brien <mob@cpan.org>},
    VERSION_FROM        => 'lib/Devel/DumpTrace.pm',
    ABSTRACT_FROM       => 'lib/Devel/DumpTrace.pm',
    (eval $ExtUtils::MakeMaker::VERSION >= 6.3002
      ? ('LICENSE'=> 'perl')
      : ()),
    (eval $ExtUtils::MakeMaker::VERSION >= 6.46
      ? ('META_MERGE'   => { no_index => $NO_INDEX })
      : ()),
    PL_FILES            => {},
    PREREQ_PM => {
        'Test::More' => 0,
	'Hash::SafeKeys' => '0.01',   # RT#77673
	'PadWalker' => 1.5,
        'Scalar::Util' => '1.14',     # 1.14 has improved reftype()
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Devel-DumpTrace-*' },
    test                => { TESTS => $TESTS },
);
